stingray/io.py

```diff
@@ -714,6 +722,516 @@ def __init__(self):
         pass
 
 
+class FITSTimeseriesReader(object):
+    main_array_attr = "time"
+
+    def __init__(
+        self,
+        fname,
+        output_class=None,
+        force_hduname=None,
+        gti_file=None,
+        gtistring=None,
+        additional_columns=None,
+        data_kind="events",
+    ):
+        self.fname = fname
+        self._meta_attrs = []
+        self.gtistring = gtistring
+        self.output_class = output_class
+        self.additional_columns = additional_columns
+        if "EventList" in str(output_class) or data_kind.lower() in ["events", "times"]:
+            self._initialize_header_events(fname, force_hduname=force_hduname)
+        else:
+            raise NotImplementedError(
+                "Only events are supported by FITSTimeseriesReader at the moment. "
+                f"{data_kind} is an unknown data kind."
+            )
+        self.data_kind = data_kind
+        if additional_columns is None and self.detector_key != "NONE":
+            additional_columns = [self.detector_key]
+        elif self.detector_key != "NONE":
+            additional_columns.append(self.detector_key)
+        self.data_hdu = fits.open(self.fname)[self.hduname]
+        self.gti_file = gti_file
+        self._read_gtis(self.gti_file)
+
+    @property
+    def time(self):
+        return self[:].time
+
+    def meta_attrs(self):
+        return self._meta_attrs
+
+    def _add_meta_attr(self, name, value):
+        """Add a meta attribute to the object."""
+        if name not in self._meta_attrs:
+            self._meta_attrs.append(name)
+        setattr(self, name, value)
+
+    @property
+    def exposure(self):
+        """
+        Return the total exposure of the time series, i.e. the sum of the GTIs.
+
+        Returns
+        -------
+        total_exposure : float
+            The total exposure of the time series, in seconds.
+        """
+
+        return get_total_gti_length(self.gti)
+
+    def __getitem__(self, index):
+        """Return an element or a slice of the object, e.g. ``ts[1]`` or ``ts[1:2]."""
+
+        data = self.data_hdu.data[index]
+
+        return self.transform_slice(data)
+
+    def transform_slice(self, data):
+        # Here there will be some logic to understand whether transfomring to events or something else
+
+        if self.data_kind == "times":
+            return data[self.time_column][:] + self.timezero
+        if self.output_class is None:
+            return data
+        if self.data_kind == "events":
+            return self._transform_slice_into_events(data)
+
+    def _transform_slice_into_events(self, data):
+        columns = [self.time_column]
```

Comment:
This method is complex enough that I think it needs a docstring as well

Reply:
Done